File: work/MuonCorrection/muonCorrection.py
import ROOT
from ROOT import RDataFrame
import numpy as np
import gc
from tqdm import tqdm
import concurrent.futures

# ROOT 설정
ROOT.EnableThreadSafety()
ROOT.EnableImplicitMT()

gc.collect()
ROOT.gSystem.Load("libAnalysisClasses.so")

# 파일 로드
mc_file = "/data1/users/dndus0107/AnalysisResults/processed_data/DYto2Mu_MLL-50to120_22EEDR.root"
data_file = "/data1/users/dndus0107/AnalysisResults/processed_data/DYto2Mu_MLL-50to120_22EEDR.root"

# RDataFrame 초기화
mc_df = RDataFrame("AnalysisMC/Analysis", mc_file)
data_df = RDataFrame("AnalysisMC/Analysis", data_file)

# 데이터 프레임 복사본 생성
data_df_copy = data_df

# Binning 설정
eta_bins = [0, 0.9, 1.2, 2.1, 2.4]
phi_bins = np.linspace(0, np.pi, 11)
charge_bins = [-1, 1]
fine_tune_factor = {}

# 초기 fine_tune_factor 설정
for eta_bin in range(len(eta_bins) - 1):
    for phi_bin in range(len(phi_bins) - 1):
        for charge in charge_bins:
            fine_tune_factor[(charge, eta_bin, phi_bin)] = 1.0

# Binning 함수 정의
def get_bin(value, bins):
    for i in range(len(bins) - 1):
        if bins[i] <= value < bins[i + 1]:
            return i
    return -1

# Z mass 평균 계산 함수 수정
def calculate_inverse_pt_average(df, charge, eta_bin, phi_bin, is_gen=False):
    df_copy = df
    muon_type = "genMuons" if is_gen else "muons"
    
    eta_min, eta_max = eta_bins[eta_bin], eta_bins[eta_bin + 1]
    phi_min, phi_max = phi_bins[phi_bin], phi_bins[phi_bin + 1]
    
    try:
        # 필터 조건을 Define으로 먼저 생성
        filtered_df = df_copy

        # 각 조건에 대한 마스크 생성
        if not is_gen:
            filtered_df = filtered_df.Define("mask_tight", f"{muon_type}.isTightReco && {muon_type}.isTightRecoZ")
        
        filtered_df = filtered_df.Define(
            "selection_mask",
            f"""
            ROOT::VecOps::RVec<bool> mask;
            for (size_t i = 0; i < {muon_type}.pt.size(); ++i) {{
                bool pass = {muon_type}.charge[i] == {charge} &&
                            std::abs({muon_type}.eta[i]) >= {eta_min} &&
                            std::abs({muon_type}.eta[i]) < {eta_max} &&
                            std::abs({muon_type}.phi[i]) >= {phi_min} &&
                            std::abs({muon_type}.phi[i]) < {phi_max};
                if (!is_gen) {{
                    pass = pass && mask_tight[i];
                }}
                mask.push_back(pass);
            }}
            return mask;
            """
        )
        
        # 선택된 뮤온의 1/pt 계산
        filtered_df = filtered_df.Define(
            "selected_inv_pt",
            f"""
            ROOT::VecOps::RVec<double> result;
            for (size_t i = 0; i < {muon_type}.pt.size(); ++i) {{
                if (selection_mask[i]) {{
                    result.push_back(1.0 / {muon_type}.pt[i]);
                }}
            }}
            return result;
            """
        )
        
        # 선택된 뮤온이 있는지 확인
        count = filtered_df.Define("n_selected", "Sum(selection_mask)").Max("n_selected").GetValue()
        
        if count == 0:
            print(f"Warning: No muons found for charge={charge}, eta_bin={eta_bin}, phi_bin={phi_bin}")
            return 0.0
        
        # 평균 계산
        result = filtered_df.Mean("selected_inv_pt").GetValue()
        print(f"Found {count} muons, average inverse pt: {result}")
        return result
        
    except Exception as e:
        print(f"Error in calculate_inverse_pt_average: {e}")
        return 0.0
#test
print(calculate_inverse_pt_average(data_df, -1, 1, 1, True))
print(calculate_inverse_pt_average(data_df, 1, 1, 1))

def calculate_inverse_pt_average_parallel(df, charge, eta_bin, phi_bin, is_gen=False):
    # 기존의 calculate_inverse_pt_average 함수를 호출
    return calculate_inverse_pt_average(df, charge, eta_bin, phi_bin, is_gen)

# Correction factors 계산
def calculate_correction_factors_parallel():
    correction_factor_Cs = {}
    correction_factor_D_ms = {}
    correction_factor_D_as = {}
    correction_factor_As = {}
    correction_factor_Ms = {}
    data_inv_pt_avg = {}

    with concurrent.futures.ThreadPoolExecutor() as executor:
        futures = {}
        for eta_bin in range(len(eta_bins) - 1):
            for phi_bin in range(len(phi_bins) - 1):
                for charge in charge_bins:
                    futures[(charge, eta_bin, phi_bin)] = executor.submit(
                        calculate_inverse_pt_average_parallel, mc_df, charge, eta_bin, phi_bin, True
                    )

        for key, future in futures.items():
            charge, eta_bin, phi_bin = key
            mc_inv_pt_avg = future.result()
            data_inv_pt_avg[key] = fine_tune_factor[key] * calculate_inverse_pt_average(data_df, charge, eta_bin, phi_bin)
            correction_factor_Cs[key] = mc_inv_pt_avg - data_inv_pt_avg[key]

        for eta_bin in range(len(eta_bins) - 1):
            for phi_bin in range(len(phi_bins) - 1):
                correction_factor_D_ms[(eta_bin, phi_bin)] = (
                    correction_factor_Cs[(1, eta_bin, phi_bin)] + correction_factor_Cs[(-1, eta_bin, phi_bin)]) / 2
                
                correction_factor_D_as[(eta_bin, phi_bin)] = (
                    correction_factor_Cs[(1, eta_bin, phi_bin)] - correction_factor_Cs[(-1, eta_bin, phi_bin)]) / 2
                
                # 분모가 0인지 확인
                denominator = data_inv_pt_avg[(1, eta_bin, phi_bin)] + data_inv_pt_avg[(-1, eta_bin, phi_bin)]
                if denominator != 0:
                    correction_factor_As[(eta_bin, phi_bin)] = (
                        correction_factor_D_as[(eta_bin, phi_bin)] -
                        ((correction_factor_D_ms[(eta_bin, phi_bin)] * (data_inv_pt_avg[(1, eta_bin, phi_bin)] - data_inv_pt_avg[(-1, eta_bin, phi_bin)])) /
                         denominator))
                    
                    correction_factor_Ms[(eta_bin, phi_bin)] = 1 + (
                        2 * correction_factor_D_ms[(eta_bin, phi_bin)] / denominator
                    )
                else:
                    # 분모가 0인 경우 기본값 설정
                    print(f"eta_bin: {eta_bin}, phi_bin: {phi_bin}에 기본팩터를 적용합니다.")
                    correction_factor_As[(eta_bin, phi_bin)] = 0
                    correction_factor_Ms[(eta_bin, phi_bin)] = 1

    return correction_factor_As, correction_factor_Ms


# Fine-tuning factor 업데이트
def update_fine_tune_factors(df):
    new_factors = {}
    for eta_bin in range(len(eta_bins) - 1):
        for phi_bin in range(len(phi_bins) - 1):
            for charge in charge_bins:
                zmass, dzmass = calculate_z_mass_stats(df, charge, eta_bin, phi_bin)
                if zmass != 0:  # 0으로 나누기 방지
                    new_factors[(charge, eta_bin, phi_bin)] = 1 + (2 * dzmass / zmass)
                else:
                    new_factors[(charge, eta_bin, phi_bin)] = 1.0
    return new_factors


# Fine-tuning 루프
max_iterations = 10
convergence_threshold = 0.001
iteration = 0
convergence = False

try :
    while not convergence and iteration < max_iterations:
        print(f"Iteration {iteration + 1}")
        
        # 현재 correction factors 계산
        correction_factors = calculate_correction_factors_parallel()
        print(f"{correction_factors}")
        
        # PT 보정 적용 및 p4 업데이트는 루프의 마지막에만 원본 데이터 프레임에 적용
        data_df = data_df.Define("corrected_pt",
            """
            ROOT::RVec<double> result;
            for(size_t i = 0; i < muons.pt.size(); ++i) {
                result.push_back(apply_correction(muons.pt[i], muons.charge[i], muons.eta[i], muons.phi[i], correction_factors));
            }
            return result;
            """)
        
        # 보정된 pt로 새로운 XYZTLorentzVector 좌표 계산
        data_df = data_df.Define("new_coordinates",
            """
            struct Coordinates {
                ROOT::RVec<double> x, y, z, t;
            };
            Coordinates result;
            for(size_t i = 0; i < muons.pt.size(); ++i) {
                double pt = corrected_pt[i];
                double eta = muons.eta[i];
                double phi = muons.phi[i];
                double p = pt * std::cosh(eta);
                result.x.push_back(pt * std::cos(phi));
                result.y.push_back(pt * std::sin(phi));
                result.z.push_back(pt * std::sinh(eta));
                result.t.push_back(std::sqrt(p*p + 0.105658*0.105658));
            }
            return result;
            """)
        
        # 기존 브랜치 덮어쓰기
        data_df = data_df.Redefine("muons.pt", "corrected_pt")
        data_df = data_df.Redefine("muons.p4.fCoordinates.fX", "new_coordinates.x")
        data_df = data_df.Redefine("muons.p4.fCoordinates.fY", "new_coordinates.y")
        data_df = data_df.Redefine("muons.p4.fCoordinates.fZ", "new_coordinates.z")
        data_df = data_df.Redefine("muons.p4.fCoordinates.fT", "new_coordinates.t")
        
        # Z mass 재계산
        data_df = data_df.Define("new_z_mass",
            """
            ROOT::RVec<double> result(zBoson.mass.size());
            ROOT::Math::XYZTLorentzVector mu1(muons.p4.fCoordinates.fX[0], 
                                            muons.p4.fCoordinates.fY[0],
                                            muons.p4.fCoordinates.fZ[0],
                                            muons.p4.fCoordinates.fT[0]);
            ROOT::Math::XYZTLorentzVector mu2(muons.p4.fCoordinates.fX[1],
                                            muons.p4.fCoordinates.fY[1],
                                            muons.p4.fCoordinates.fZ[1],
                                            muons.p4.fCoordinates.fT[1]);
            result[1] = (mu1 + mu2).M();
            return result;
            """)
        data_df = data_df.Redefine("zBoson.mass", "new_z_mass")
        
        # Fine-tune factors 업데이트
        previous_factors = fine_tune_factor.copy()
        fine_tune_factor = update_fine_tune_factors(data_df)
        
        # 수렴 확인
        max_diff = 0.0
        for key in fine_tune_factor:
            diff = abs(fine_tune_factor[key] - previous_factors[key])
            max_diff = max(max_diff, diff)
        
        convergence = max_diff < convergence_threshold
        iteration += 1
        
        print(f"Maximum difference: {max_diff}")
except KeyboardInterrupt :
    print("교정을 멈추고 현재 상태를 저장합니다.")

# 최종 결과 저장
output_file_data = "data_corrected.root"
data_df.Snapshot("Analysis/Analysis", output_file_data)
print(f"Data 교정 파일 저장: {output_file_data}")

File: work/MuonCorrection/legacy/muonCorrection.cc
#include <TFile.h>
#include <TTree.h>
#include <TLorentzVector.h>
#include <vector>
#include <iostream>
#include <map>
#include <string>
#include <cmath>
#include <memory>
#include <tuple>
#include "interface/AnalysisClasses.h"
#include <array>
#include "ROOT/RDataFrame.hxx"
#include <TSystem.h>
class MuonAnalyzer {
private:
    // Binning parameters
    std::vector<double> eta_bins;
    std::vector<double> phi_bins;
    std::vector<int> charge_bins;
    
    // Fine tuning factors
    std::map<std::tuple<int, int, int>, double> fine_tune_factors;
    
    // Constants
    const double Z_MASS = 91.1876;
    const double MAX_ITERATIONS = 10;
    const double CONVERGENCE_THRESHOLD = 0.001;

    // Helper function to get bin index
    int getBin(double value, const std::vector<double>& bins) {
        for (size_t i = 0; i < bins.size() - 1; ++i) {
            if (value >= bins[i] && value < bins[i + 1]) return i;
        }
        return -1;
    }

    // Calculate Z mass statistics for a given bin
    std::pair<double, double> calculateZMassStats(TTree* tree, int charge, int eta_bin, int phi_bin) {
        std::vector<double>* muon_pt = nullptr;
        std::vector<double>* muon_eta = nullptr;
        std::vector<double>* muon_phi = nullptr;
        std::vector<int>* muon_charge = nullptr;
        std::vector<bool>* muon_isTightReco = nullptr;
        std::vector<bool>* muon_isTightRecoZ = nullptr;
        std::vector<double>* z_mass = nullptr;

        tree->SetBranchAddress("muons.pt", &muon_pt);
        tree->SetBranchAddress("muons.eta", &muon_eta);
        tree->SetBranchAddress("muons.phi", &muon_phi);
        tree->SetBranchAddress("muons.charge", &muon_charge);
        tree->SetBranchAddress("muons.isTightReco", &muon_isTightReco);
        tree->SetBranchAddress("muons.isTightRecoZ", &muon_isTightRecoZ);
        tree->SetBranchAddress("zBoson.mass", &z_mass);

        double sum_mass = 0;
        double sum_dmass = 0;
        int count = 0;

        Long64_t nentries = tree->GetEntries();
        for (Long64_t i = 0; i < nentries; ++i) {
            tree->GetEntry(i);

            for (size_t j = 0; j < muon_pt->size(); ++j) {
                if (!(*muon_isTightReco)[j] || !(*muon_isTightRecoZ)[j]) continue;
                if ((*muon_charge)[j] != charge) continue;

                double eta = std::abs((*muon_eta)[j]);
                double phi = std::abs((*muon_phi)[j]);

                if (getBin(eta, eta_bins) == eta_bin && getBin(phi, phi_bins) == phi_bin) {
                    sum_mass += (*z_mass)[1];  // Using index 1 for tightReco Z mass
                    sum_dmass += (*z_mass)[1] - Z_MASS;
                    count++;
                }
            }
        }

        if (count == 0) return {0, 0};
        return {sum_mass / count, sum_dmass / count};
    }

    // Calculate inverse pt average
    double calculateInversePtAverage(TTree* tree, int charge, int eta_bin, int phi_bin, bool is_gen = false) {

        double sum_inv_pt = 0;
        int nMuon = 0;

        if (is_gen) {
            std::vector<GenMuonInfo>* genMuons = nullptr;
            tree->SetBranchAddress("genMuons", &genMuons);
            Long64_t nentries = tree->GetEntries();
            for (Long64_t i = 0; i < nentries; ++i) {
                tree->GetEntry(i);
                for (const auto& muon : *genMuons) {
                    if (!(muon.isGenZ)) continue;
                    if ((muon.charge) != charge) continue;

                    double eta = std::abs((muon.eta));
                    double phi = std::abs((muon.phi));

                    if (getBin(eta, eta_bins) == eta_bin && getBin(phi, phi_bins) == phi_bin && muon.pt > 0) {
                        sum_inv_pt += 1.0 / (muon.pt);
                        nMuon++;
                    }
                }
            }
        }
        else {
            std::vector<SelectedMuon>* muons = nullptr;            
            tree->SetBranchAddress("muons", &muons);
            Long64_t nentries = tree->GetEntries();
            for (Long64_t i = 0; i < nentries; ++i) {
                tree->GetEntry(i);
                for (const auto& muon : *muons) {
                    if (!(muon.isTightRecoZ)) continue;
                    if ((muon.charge) != charge) continue;
                    double eta = std::abs((muon.eta));
                    double phi = std::abs((muon.phi));

                    if (getBin(eta, eta_bins) == eta_bin && getBin(phi, phi_bins) == phi_bin) {
                        sum_inv_pt += 1.0 / (muon.pt);
                        nMuon++;
                    }
                }
            }
        }


        std::cout << "calculateInversePtAverage: "<< sum_inv_pt / nMuon << std::endl;

        if (nMuon == 0) return 0;
        return sum_inv_pt / nMuon;
    }

    // Calculate correction factors
    struct CorrectionFactors {
        std::map<std::tuple<int, int, int>, double> Cs;
        std::map<std::tuple<int, int>, double> D_ms;
        std::map<std::tuple<int, int>, double> D_as;
        std::map<std::tuple<int, int>, double> As;
        std::map<std::tuple<int, int>, double> Ms;
    };

    CorrectionFactors calculateCorrectionFactors(TTree* mc_tree, TTree* data_tree) {
        CorrectionFactors factors;
        std::map<std::tuple<int, int, int>, double> data_inv_pt_avg;

        // Calculate Cs factors
        for (size_t eta_bin = 0; eta_bin < eta_bins.size() - 1; ++eta_bin) {
            for (size_t phi_bin = 0; phi_bin < phi_bins.size() - 1; ++phi_bin) {
                for (int charge : charge_bins) {
                    auto key = std::make_tuple(charge, eta_bin, phi_bin);
                    double mc_inv_pt = calculateInversePtAverage(mc_tree, charge, eta_bin, phi_bin, true);
                    data_inv_pt_avg[key] = fine_tune_factors[key] * 
                                         calculateInversePtAverage(data_tree, charge, eta_bin, phi_bin);
                    factors.Cs[key] = mc_inv_pt - data_inv_pt_avg[key];
                }
            }
        }

        // Calculate D_ms, D_as, As, and Ms factors
        for (size_t eta_bin = 0; eta_bin < eta_bins.size() - 1; ++eta_bin) {
            for (size_t phi_bin = 0; phi_bin < phi_bins.size() - 1; ++phi_bin) {
                auto pos_key = std::make_tuple(1, eta_bin, phi_bin);
                auto neg_key = std::make_tuple(-1, eta_bin, phi_bin);
                auto bin_key = std::make_tuple(eta_bin, phi_bin);

                factors.D_ms[bin_key] = (factors.Cs[pos_key] + factors.Cs[neg_key]) / 2.0;
                factors.D_as[bin_key] = (factors.Cs[pos_key] - factors.Cs[neg_key]) / 2.0;

                double denominator = data_inv_pt_avg[pos_key] + data_inv_pt_avg[neg_key];
                if (denominator != 0) {
                    factors.As[bin_key] = factors.D_as[bin_key] - 
                        ((factors.D_ms[bin_key] * (data_inv_pt_avg[pos_key] - data_inv_pt_avg[neg_key])) / denominator);
                    factors.Ms[bin_key] = 1 + (2 * factors.D_ms[bin_key] / denominator);
                } else {
                    std::cout << "Warning: Zero denominator for eta_bin=" << eta_bin 
                              << ", phi_bin=" << phi_bin << std::endl;
                    factors.As[bin_key] = 0;
                    factors.Ms[bin_key] = 1;
                }
            }
        }

        return factors;
    }

    // Update fine tuning factors
    void updateFineTuneFactors(TTree* tree) {
        for (size_t eta_bin = 0; eta_bin < eta_bins.size() - 1; ++eta_bin) {
            for (size_t phi_bin = 0; phi_bin < phi_bins.size() - 1; ++phi_bin) {
                for (int charge : charge_bins) {
                    auto [zmass, dzmass] = calculateZMassStats(tree, charge, eta_bin, phi_bin);
                    if (zmass != 0) {
                        auto key = std::make_tuple(charge, eta_bin, phi_bin);
                        fine_tune_factors[key] = 1 + (2 * dzmass / zmass);
                    }
                }
            }
        }
    }

public:
    MuonAnalyzer() {
        // Initialize binning
        eta_bins = {0, 0.9, 1.2, 2.1, 2.4};
        phi_bins.resize(11);
        double phi_step = M_PI / 10;
        for (int i = 0; i < 11; ++i) {
            phi_bins[i] = i * phi_step;
        }
        charge_bins = {-1, 1};

        // Initialize fine tuning factors
        for (size_t i = 0; i < eta_bins.size() - 1; ++i) {
            for (size_t j = 0; j < phi_bins.size() - 1; ++j) {
                for (int charge : charge_bins) {
                    fine_tune_factors[std::make_tuple(charge, i, j)] = 1.0;
                }
            }
        }
    }

    void analyze(const std::string& mc_file_path, const std::string& data_file_path) {
        // Open files
        std::unique_ptr<TFile> mc_file(TFile::Open(mc_file_path.c_str(), "READ"));
        std::unique_ptr<TFile> data_file(TFile::Open(data_file_path.c_str(), "READ"));

        if (!mc_file || !data_file) {
            std::cerr << "Error opening files" << std::endl;
            return;
        }

        TTree* mc_tree = (TTree*)mc_file->Get("AnalysisMC/Analysis");
        //TTree* data_tree = (TTree*)data_file->Get("AnalysisMC/Analysis");

        //TTree* mc_tree = (TTree*)mc_file->Get("Analysis");
        TTree* data_tree = (TTree*)data_file->Get("Analysis");

        if (!mc_tree || !data_tree) {
            std::cerr << "Error accessing trees" << std::endl;
            return;
        }

        // Perform iterations
        int iteration = 0;
        bool convergence = false;

        try {
            while (!convergence && iteration < MAX_ITERATIONS) {
                std::cout << "Iteration " << (iteration + 1) << std::endl;

                // Calculate correction factors
                auto correction_factors = calculateCorrectionFactors(mc_tree, data_tree);

                // Store previous factors for convergence check
                auto previous_factors = fine_tune_factors;

                // Update fine tune factors
                updateFineTuneFactors(data_tree);
                
                double sum_squared_diff = 0.0;
                int total_factors = 0;
                // Check convergence 
                for (const auto& [key, value] : fine_tune_factors) {
                    double diff_from_one = value - 1.0;
                    sum_squared_diff += diff_from_one * diff_from_one;
                    total_factors++;
                }
                
                double mean_squared_diff = sum_squared_diff / total_factors;
                std::cout << "Mean squared difference from 1.0: " << mean_squared_diff << std::endl;
                
                // 수렴 여부 확인
                convergence = mean_squared_diff < CONVERGENCE_THRESHOLD;
                iteration++;
                
                if (convergence) {
                    std::cout << "Convergence achieved after " << iteration << " iterations" << std::endl;
                }
    
            }
            
            if (!convergence) {
                std::cout << "Maximum iterations reached without convergence" << std::endl;
            }
        }
        catch (const std::exception& e) {
            std::cerr << "Error during analysis: " << e.what() << std::endl;
        }

        // Save results
        std::unique_ptr<TFile> output_file(TFile::Open("data_corrected.root", "RECREATE"));
        if (output_file) {
            output_file->cd();
            // Here you would write the corrected data
            // This part would need to be implemented based on your specific needs
            std::cout << "Results saved to data_corrected.root" << std::endl;
        }
    }
};


int muonCorrection() {
    gSystem->Load("libAnalysisClasses.so");
    MuonAnalyzer analyzer;
    analyzer.analyze(
        "/data1/users/dndus0107/AnalysisResults/processed_data/DYto2Mu_MLL-50to120_22EEDR.root", //MC
        "/data1/users/dndus0107/AnalysisResults/processed_data/mini_Analysis_Data_22EFG.root" //DATA
    );
    return 0;
}